robotbase/src/main/java/com/team2357/frc2022/subsystems/TurretSubsystem.java

```diff
@@ -0,0 +1,241 @@
+        resetHeading();
+    }
+
+    public void configure(Configuration config) {
+        m_config = config;
+
+        m_turretMotor.setIdleMode(m_config.turretMotorIdleMode);
+        m_turretMotor.setSmartCurrentLimit(m_config.m_turretMotorStallLimitAmps, m_config.m_turretMotorFreeLimitAmps);
+
+        m_pidController = m_turretMotor.getPIDController();
+
+        // PID coefficients
+        m_config.m_turretMotorP = 5e-5;
+        m_config.m_turretMotorI = 1e-6;
+        m_config.m_turretMotorD = 0;
+        m_config.m_turretMotorIZone = 0;
+        m_config.m_turretMotorFF = 0.000156;
+        m_config.m_turretMotorMinOutput = 1;
+        m_config.m_turretMotorMinOutput = -1;
+        m_config.m_turretMotorMaxRPM = 5700;
+
+        // Smart Motion Coefficients
+        m_config.m_turretMotorMaxVel = 2000; // rpm
+        m_config.m_turretMotorMaxAcc = 1500;
+
+        // set PID coefficients
+        m_pidController.setP(m_config.m_turretMotorP);
+        m_pidController.setI(m_config.m_turretMotorI);
+        m_pidController.setD(m_config.m_turretMotorD);
+        m_pidController.setIZone(m_config.m_turretMotorIZone);
+        m_pidController.setFF(m_config.m_turretMotorFF);
+        m_pidController.setOutputRange(m_config.m_turretMotorMinOutput, m_config.m_turretMotorMinOutput);
+
+        // Configure smart motion
+        int smartMotionSlot = 0;
+        m_pidController.setSmartMotionMaxVelocity(m_config.m_turretMotorMaxVel, smartMotionSlot);
+        m_pidController.setSmartMotionMinOutputVelocity(m_config.m_turretMotorMinVel, smartMotionSlot);
+        m_pidController.setSmartMotionMaxAccel(m_config.m_turretMotorMaxAcc, smartMotionSlot);
+        m_pidController.setSmartMotionAllowedClosedLoopError(m_config.m_turretMotorAllowedError, smartMotionSlot);
+    }
+
+    public void setTurretSpeed(double speed) {
+        m_turretMotor.set(speed);
+    }
+
+    public void resetHeading() {
+        m_turretMotor.getEncoder().setPosition(0);
+    }
+
+    public double getTurretRotations() {
+        return m_turretMotor.getEncoder().getPosition();
+    }
+
+    public void setTurretPosition(double position) {
+        m_setPoint = position;
+        m_pidController.setReference(m_setPoint, CANSparkMax.ControlType.kSmartMotion);
+    }
+
+    public boolean atSetPoint() {
+        return com.team2357.frc2022.util.Utility.isWithinTolerance(getTurretRotations(), m_setPoint,
+                m_config.m_turretMotorAllowedError);
+    }
+
+    public boolean atSetPoint(double setPoint) {
+        return com.team2357.frc2022.util.Utility.isWithinTolerance(getTurretRotations(), setPoint,
+                m_config.m_turretMotorAllowedError);
+    }
+
+    public boolean isOnZero() {
+        boolean isMagnetDetected = false;
+
+        if (m_arduinoHallEffectSensor.isConnected()) {
+            isMagnetDetected = !m_arduinoHallEffectSensor
+                    .getDeviceFieldBoolean(Constants.ARDUINO.TURRET_HALL_SENSOR_NAME, "state");
```

Comment:
This will become cleaner when the arduino details are in their own class. I really don't like the turret knowing anything about an arduino implementation.